### PR TITLE
Add support for AWS dynamic discovery of nodes in a memcached cluster

### PR DESCRIPTION
**Pull request by jdolinski**
[AWS AutoDiscovery](http://docs.aws.amazon.com/AmazonElastiCache/latest/UserGuide/AutoDiscovery.html) 
AWS Elasticache Memcached support allows auto-discovery to dynamically manage memcached nodes in the cluster. Amazon ElastiCache Cluster Client is an enhanced PHP library to connect to ElastiCache clusters. This client library has been built upon libmemcached and is released under the Apache License. https://github.com/awslabs/aws-elasticache-cluster-client-memcached-for-php/releases

Current Joomla release as of today (3.6.5) does not support AWS's autodiscovery of nodes and design is based on configuring static nodes in a multi memcached cluster. As of today, the static feature is not fully implemented and will require you to update your configuration if new nodes or failures occur in your cluster. This is a new feature request to support PHP autodiscovery.
Pull Request for Issue #14993.

### Summary of Changes
Added support to allow [auto discovery of memcached nodes from AWS ElastiCache PHP client](https://github.com/awslabs/aws-elasticache-cluster-client-memcached-for-php). 

1. Created a subclass of the JCacheStorageMemcached for AWS support called JCacheStorageMemcachedaws.
2. Created new cache handler memcachedaws

### Testing Instructions
Prerequisite documentation for autodiscovery configuration.
[Installing the ElastiCache Cluster Client for PHP](http://docs.aws.amazon.com/AmazonElastiCache/latest/UserGuide/Appendix.PHPAutoDiscoverySetup.html).
[Compiling the Source Code for the ElastiCache Cluster Client for PHP](http://docs.aws.amazon.com/AmazonElastiCache/latest/UserGuide/Appendix.PHPAutoDiscoveryCompile.html)

Configure your configuration.php as follows:

public cache_handler = 'memcachedaws'
...
public $memcached_persist = '1';
public $memcached_compress = '1';
public $memcached_server_host = 'your-elasticache-endpoint';
public $memcached_server_port = '11211';
public $memcached_autodiscovery = '1';
...
public session_handler = 'memcached'
public $session_memcached_server_host = 'your-elasticache-endpoint';
public $session_memcached_server_port = '11211';

The configuration.php file will be required to be modified to include the following line:
  **public $memcached_autodiscovery = '1';**
This will allow you to turn on the auto discovery to register all the memcached nodes in your cluster with the Joomla Memcached.

### Expected result
Using Auto Discovery Caching with multiple Memcached nodes. You can login and you will not be logged out before the session expiration as you are routed to the nodes in your cluster.

When turning off auto discovery the caching behavior remains same as before this change.

### Actual result
Caching in memcached & session management hits.

### Documentation Changes Required
I have updated the "administrator/components/com_config/model/form/application.xml"

### CODE REVIEW

File: administrator/language/en-GB/en-GB.com_config.ini [hunk 1]
```diff
@@ -139,6 +139,8 @@ COM_CONFIG_FIELD_MAIL_SMTP_USERNAME_DESC="Enter the username for access to the S
 COM_CONFIG_FIELD_MAIL_SMTP_USERNAME_LABEL="SMTP Username"
 COM_CONFIG_FIELD_MEMCACHE_COMPRESSION_DESC="Memcache(d) compression."
 COM_CONFIG_FIELD_MEMCACHE_COMPRESSION_LABEL="Memcache(d) Compression"
+COM_CONFIG_FIELD_MEMCACHE_AUTODISCOVERY_DESC="Use AWS autodiscovery features, requires AWS ElasticCache Cluster Client installed."
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_config.ini]: It is our practice that all language strings are alphabetically ordered. So these two new strings need to be moved up two rows

**Comment by brianteeman** [comment]
As you can see the travis continuous integration tests are failing with

`Fatal error: Undefined class constant 'OPT_CLIENT_MODE' in /home/travis/build/joomla/joomla-cms/libraries/joomla/cache/storage/memcached.php on line 88`

**Comment by jdolinski** [comment]
@brianteeman Thanks sorry I didn't realize I opened two issues. 

The Travis CI failed because of the dependency on AWS memcached module. Do you need me to work through that? 

This made me realize I also need to test this pull on a server using the standard PHP memcached module. I will do this tomorrow.

**Comment by brianteeman** [comment]
@mbabker can you comment on that last question as its above my paygrade

**Comment by mbabker** [comment]
We might have to do this as a subclass of the Memcached adapter versus including it inline if it has an additional dependency.  That's my first thought.

**Comment by jdolinski** [comment]
I can subclass the memcached class, if that is preferred. It is a different implementation of the memcached module. Let me know your preference and any tips.

**Comment by jdolinski** [comment]
BTW - I stumbled upon this issue when trying to use a memcached multi-node cluster with Session Management. The constructor in the [memcached.php](https://github.com/joomla/joomla-cms/blob/staging/libraries/joomla/session/storage/memcached.php) file had the following //todo. This can be accomplished using the AWS memcached module on any PHP server even outside AWS. 
`		// This will be an array of loveliness
		// @todo: multiple servers
		$this->_servers = array(
			array(
				'host' => $config->get('session_memcached_server_host', 'localhost'),
				'port' => $config->get('session_memcached_server_port', 11211),
			),
		);
`
Looking forward to your feedback.

**Comment by csthomas** [comment]
IMO subclass will be better. 
There is a method `isSupported()` which has to be overrided.

**Comment by jdolinski** [comment]
Added subclass for review and updated ticket documentation for setting up AWS ElastiCache. 
Will finish testing on my side next week. Going to create a Apache JMeter load script to test performance and dynamically adding/removing nodes.

**Comment by jdolinski** [comment]
I did not override isSupported() AWS uses same Memcached class

**Comment by zero-24** [comment]
@jdolinski i have just commited some CS Changes and extended doc blocks https://github.com/joomla/joomla-cms/pull/14993/commits/91fd32e74d3c803904cccdd037b6488c7cf6f54e

File: libraries/joomla/cache/storage/memcachedaws.php [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+        $stats  = parent::$_db->getStats();
+        $result = !empty($stats["$host:$port"]) && $stats["$host:$port"]['pid'] > 0;
+
+        if (!$result)
+        {
+            // Null out the connection to inform the constructor it will need to attempt to connect if this class is instantiated again
+            parent::$_db = null;
+
+            throw new JCacheExceptionConnecting('Could not connect to memcached server. '. parent::$_db->getResultMessage());
```
Comment (zero-24) [inline libraries/joomla/cache/storage/memcachedaws.php]: you are throwing here a exeption but you dont catch it in `getStaticClientConnection()` and `getDynamicClientConnection()` is this expected?

Comment (jdolinski) [inline libraries/joomla/cache/storage/memcachedaws.php]: I used the same pattern as the superclass. An exception was thrown but not caught.

**Comment by csthomas** [comment]
I mentioned about  `isSupported()` in order to add a condition: if `Memcached::OPT_CLIENT_MODE` is defined.

**Comment by jdolinski** [comment]
Gotcha, will override that method

**Comment by jdolinski** [comment]
I overrode the isSupported() method, please review latest changes. Will be pushing this version to production and begin using with AWS ElastiCache.

Created Apache Jmeter test and discovered bug under load, refactored the dynamic connection code. Was able to add/remove memcached nodes while simulating user connections to multiple pages in test.

File: administrator/language/en-GB/en-GB.com_config.ini (hunk 1 above)
Comment (jdolinski) [inline administrator/language/en-GB/en-GB.com_config.ini]: Fixed

**Comment by jdolinski** [comment]
@brianteeman I ran performance tests against memcached and found my site is slower. Very similar and appears to be related to #9898. I will take a look at the comments and see if I can help. Maybe this is the reason the cache handler is still marked as "Experimental"?

**Comment by HLeithner** [comment]
Hi @jdolinski,
thx for the time you invested into this PR. We can't merge it into J3 because of the feature freeze.
I'm also not sure if this is the right why to get it into J4 a plugin system for cache/session stores would be a better approach but if you are still interested to get this cache store into Joomla! it would be great if you rebase it for J4.

In the meantime I close this PR, thx for trying to make Joomla! a better CMS.

**Pull request closed without merging**